pageserver/src/feature_resolver.rs

```diff
@@ -0,0 +1,108 @@
+use std::{collections::HashMap, sync::Arc, time::Duration};
+
+use posthog_client_lite::{FeatureStore, PostHogClient, PostHogEvaluationError};
+use tokio_util::sync::CancellationToken;
+use utils::id::TenantId;
+
+use crate::config::PageServerConf;
+
+struct FeatureResolverInner {
+    posthog_client: PostHogClient,
+    feature_store: Arc<std::sync::Mutex<FeatureStore>>,
+    cancel: CancellationToken,
+}
+
+impl FeatureResolverInner {
+    fn new(conf: &'static PageServerConf, shutdown_pageserver: CancellationToken) -> Option<Self> {
+        if let Some(ref posthog_config) = conf.posthog_config {
+            let this = Self {
+                posthog_client: PostHogClient::new(
+                    posthog_config.server_api_key.clone(),
+                    posthog_config.client_api_key.clone(),
+                    posthog_config.project_id.clone(),
+                    posthog_config.private_api_url.clone(),
+                    posthog_config.public_api_url.clone(),
+                ),
+                feature_store: Arc::new(std::sync::Mutex::new(FeatureStore::new())),
+                cancel: shutdown_pageserver,
+            };
+            Some(this)
+        } else {
+            None
+        }
+    }
+
+    fn spawn(self: Arc<Self>, handle: &tokio::runtime::Handle) {
+        let this = self.clone();
```

Comment:
Shound't we wait for the first update on startup?
Otherwise, we will have a short period after restart where feature flags are not available, so even 100% roll out options will be temporary "rolled back"

Reply:
Yes, but it's up to the client how to handle this. Because it's expected that setting things up may take a while (not only fetching the PostHog spec, we also need to get plan type / resident size / logical size if it's the triggering condition).